src/gretel_trainer/benchmark/gretel/executor.py

```diff
@@ -0,0 +1,56 @@
+from typing import Callable, Union
+
+import pandas as pd
+
+from gretel_trainer.benchmark.core import DataSource
+from gretel_trainer.benchmark.gretel.models import GretelModel
+from gretel_trainer.benchmark.gretel.sdk import GretelSDK, GretelSDKExecutor
+from gretel_trainer.benchmark.gretel.trainer import GretelTrainerExecutor, Trainer
+
+
+class GretelExecutor:
+    def __init__(
+        self,
+        project_name: str,
+        model: GretelModel,
+        sdk: GretelSDK,
+        trainer_factory: Callable[..., Trainer],
+        benchmark_dir: str,
+    ):
+        self._project_name = project_name
+        self._model = model
+        self._internal_executor = _choose_executor(project_name, model, sdk, trainer_factory, benchmark_dir)
+
+    @property
+    def model_name(self) -> str:
+        return self._model.name
+
+    def runnable(self, source: DataSource) -> bool:
+        return self._model.runnable(source)
+
+    def train(self, source: str, **kwargs) -> None:
+        self._internal_executor.train(source, **kwargs)
+
+    def generate(self, **kwargs) -> pd.DataFrame:
+        return self._internal_executor.generate(**kwargs)
+
+    def get_sqs_score(self, synthetic: pd.DataFrame, reference: str) -> int:
+        return self._internal_executor.get_sqs_score(synthetic, reference)
+
+
+def _choose_executor(
+    project_name: str,
+    model: GretelModel,
+    sdk: GretelSDK,
+    trainer_factory: Callable[..., Trainer],
+    benchmark_dir: str,
+) -> Union[GretelTrainerExecutor, GretelSDKExecutor]:
```

Comment:
Could this return be hinted just as `Trainer`? Which is the protocol that you defined.

Reply:
This is a really interesting question.

I don't think hinting the return value as `Trainer` is appropriate here because that protocol is designed to capture the contract Benchmark has with the actual `gretel_trainer.Trainer` class. The fact that it is currently the exact same interface is (somewhat) incidental, and more importantly may not always be the case. For example, as Evaluate evolves to provide more info about synthetic datasets, we may want the "internal executors" (`GretelTrainerExecutor` and `GretelSDKExecutor`) to expose a single method—something like `get_synthetic_evaluation`—that under the hood may may require the GretelTrainerExecutor to aggregate results of a few Trainer methods (maybe something like `get_sqs_score` + `get_some_other_info` + `get_a_third_aspect`). I would add those low-level operations to the `Trainer` protocol to indicate "here's what I-the-client-Benchmark expect to be able to do with Trainer-the-library" / "these are the public functions from Trainer that I'm using". (Also worth noting, the mechanism for fetching that same enhanced Evaluate data from the SDK would likely be a little different, and so that branch of Gretel model execution would not be able to adhere to the same interface expected by this return type if it were `Trainer`.)

This question did make me wonder if there is some other protocol we could use here instead of the concrete union type, but ultimately I think the current return type hint is most appropriate. It expresses the fact that we're choosing between one of two (and only ever two) options, and those two options are really just private implementation details of the `GretelExecutor` class. To that end, I'm pushing a commit that makes that even more clear by bringing those classes into the same module. I think it actually makes the purposes of _three_ modules (executor, trainer, and sdk) a bit more clear—let me know if you agree!

Reply:
I see, thank you for all the context on this!
> It expresses the fact that we're choosing between one of two (and only ever two) options, and those two options are really just private implementation details of the `GretelExecutor` class.

This makes perfect sense, thanks!